crates/sui-upload-display/src/main.rs

```diff
@@ -0,0 +1,576 @@
+
+        let object_type = parse_hex(&record[0])?;
+        let display_id = parse_hex(&record[1])?;
+        let display_version = record[2]
+            .parse::<i16>()
+            .context("Failed to parse display version")?;
+        let display = parse_hex(&record[3])?;
+
+        let entry = DisplayEntry {
+            object_type,
+            display_id: display_id.clone(),
+            display_version,
+            display,
+        };
+        display_data.displays.insert(display_id, entry);
+    }
+
+    info!(
+        "Loaded {} display entries from file",
+        display_data.displays.len()
+    );
+    Ok(display_data)
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let _guard = TelemetryConfig::new().with_env().init();
+    info!("Starting sui-upload-display service");
+    let config = Config::parse();
+
+    let remote_store: Option<Arc<dyn ObjectStore>> =
+        match (&config.gcs_cred_path, &config.gcs_display_bucket) {
+            (Some(cred_path), Some(bucket)) => {
+                info!(
+                    "Initializing GCS with bucket '{}' using credentials from '{}'",
+                    bucket, cred_path
+                );
+
+                let gcs_builder = GoogleCloudStorageBuilder::new()
+                    .with_service_account_path(cred_path.clone())
+                    .with_bucket_name(bucket.clone());
+
+                match gcs_builder.build() {
+                    Ok(store) => {
+                        info!("Successfully initialized GCS client");
+                        Some(Arc::new(store))
+                    }
+                    Err(e) => {
+                        warn!("Failed to initialize GCS client: {}", e);
+                        None
+                    }
+                }
+            }
+            _ => {
+                warn!(
+                    "Either GCS cred path or bucket is not set, display data will not be uploaded"
+                );
+                None
+            }
+        };
+
+    let registry = Registry::new();
+    let metrics = IndexerMetrics::new(&registry);
+
+    let remote_url = Url::parse(&config.remote_url).context("Failed to parse remote URL")?;
+    let client = IngestionClient::new_remote(remote_url, metrics)?;
+    info!("Initialized remote client with URL: {}", config.remote_url);
+
+    let cancellation_token = CancellationToken::new();
+    let mut accumulated_display_data = None;
+
+    let (starting_checkpoint, latest_file_info) = find_last_processed_checkpoint(&config).await?;
+    if let Some((epoch, file_path)) = latest_file_info {
+        match load_display_entries_from_file(&config, &file_path, epoch).await {
+            Ok(loaded_data) => {
+                info!(
+                    "Successfully loaded display data from previous epoch {} with {} entries",
+                    epoch,
+                    loaded_data.displays.len()
+                );
+                accumulated_display_data = Some(loaded_data);
```

Comment:
Initialise `accumulated_display_data` using an `if` expression, rather than setting it to `None` and then overwriting it here.